Implemented Visual.Children as an implicitly linked list (optimization)

Duckers opened the pull request:
This implementation avoids all memory allocation, copying and shifting in connection with manipulating the Children list, and all the list operations are now O(1), including Contains(), InsertAfter() and Remove()

Updated all relevant call sites to use this API over the IList<T> interface (which is slower).

This is a big optimization over using MiniList, shaving 2-3 seconds of a case where 5000 text elements where added to a list.

** Briefly describe changes and the motivation behind them here **

This PR contains:
- [x] Changelog
- [x] Documentation
- [-] Tests - These code paths are covered plenty by our existing tests

mortoray commented on Source/Fuse.Controls.Panels/Layouts/WrapLayout.uno:
```diff
@@ -82,21 +82,21 @@ namespace Fuse.Layouts
 		}
 		
-		float2 Arrange(IList<Node> elements, LayoutParams lp,	
+		float2 Arrange(Visual elements, LayoutParams lp,	
```
The name of the argument needs to change on these, `elements` is  misleading, especially when you see things like `elements.Children.Count` in the code.

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -29,24 +29,24 @@ namespace Fuse
 		}
 
-		public void RemoveAllChildren<T>()
+		public void RemoveAllChildren<T>() where T: Node
```
This restriction really isn't needed, `where T : class` is enough. Somebody (not us though) might be calling this with an interface that doesn't derive from `Node`.

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -29,24 +29,24 @@ namespace Fuse
-
-				i = Math.Min(i, Children.Count) - 1;
+				var v = c as T;
+				if (v != null) Children_Remove(v);
```
Doesn't this imply the loop needs `Children_SafeForEach` or equivalent?  It's removing items from the list it is iterating.

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -167,19 +165,21 @@ namespace Fuse
 		void ICollection<Node>.Clear()
 		{
-			foreach (var child in _children)
+			foreach (var child in Children)
```
Why isn't this using the linked list iteration?

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -258,18 +250,16 @@ namespace Fuse
 					{
-						var where = IndexOf(c);
-						if (where != -1)
+						if (Children_Contains(c))
```
(not about this line, about the HashSet)

I wonder if we can't assume that `InsertOrMove` is exclusive and add an intrinsic boolean flag to node to track the `moved` status done here. That would avoid the HashSet.

Duckers replied:
That would be a separate optimziation for later. I don't want the extra code churn and added field on this branch, and the HashSet doesn't show up as significant in any profiling.

mortoray commented on Source/Fuse.Nodes/Visual.Layout.uno:
```diff
@@ -329,10 +329,13 @@ namespace Fuse
-				for (int i = 0; i < Children.Count; i++)
+
+				var c = Children_first;
+				while (c != null)
```
Why did you switch to `while` here instead of the `for` loop used elsewhere?

mortoray commented on Source/Fuse.Nodes/Visual.ZOrder.uno:
```diff
@@ -14,9 +14,14 @@ namespace Fuse
 			{
-				if (!HasVisualChildren) return null;
-
-				return FirstChild<Visual>();
```
I see no reason to change this bit since `FirstChild` uses the new iteration. You're just inlining a copy of it here.

mortoray commented on Source/Fuse.Nodes/Visual.ZOrder.uno:
```diff
@@ -142,15 +147,18 @@ namespace Fuse
 
-			for (int i = 0; i < nodes.Count; i++)
+			var c = Children_first;
+			while (c != null)
```
Why not the `for` loop?

mortoray commented on Source/Fuse.Reactive.Bindings/Instance.WindowItems.uno:
```diff
@@ -175,7 +175,7 @@ namespace Fuse.Reactive
 			var lastNode = GetLastNodeFromIndex(windowIndex-1);
 
-			Parent.InsertOrMoveNodes( Parent.Children.IndexOf(lastNode) + 1, wi.Nodes.GetEnumerator() );
+			Parent.InsertOrMoveNodes( lastNode, wi.Nodes.GetEnumerator() );
```
The `InsertOrMoveNodes` name should probably be changed to `InsertOrMoveNodesAfter` to match the `InsertNodesAfter` function.

mortoray commented:
Your original idea for copy-on-modify for iteration is better. The `SafeForEach` would then look something more like this:

```
		class SafeIterator
		{
			public SafeIterator Previous;
			public Node[] List;
			public Node Current;
			public Action<Visual, Node> Action;
		}
		SafeIterator Children_CurrentIterator;

		void Children_SafeForEach(Action<Visual, Node> action)
		{
			Children_CompleteCurrentAction();
			
			var iter = new SafeIterator{ Current = Children_first,
				Action = action, 
				Previous = Children_CurrentIterator }; //we could use a simple pool if allocation is an issue
			Children_CurrentIterator = iter;
			while (iter.Current != null) 
			{
				iter.Action(iter.Current);
				//if interrupted use the copied list instead
				if (iter.List != null )
				{
					for (int i=0; i < iter.List.Length; ++i)
						iter.Action(iter.List[i]);
					break;
				}
				
				iter.Current = iter.Current.Children_next;
			}
			
			if (iterm.List == null)
				Children_CurrentIterator = iter.Previous;
		}

		void Children_CompleteCurrentAction()
		{
			var iter = Children_CurrentIterator;
			while (iter != null)
			{
				if (iter.List == null)
					iterList = Children_ToArrayAfter(iter.Current);
				
				iter = iter.Previous;
			}
			
			//no need to track current anymore
			Children_CurrentIterator = null;
		}
```

This avoids any deep stacks. It keeps the same processing order we had previously. It allows nesting of iteration.

You'd have to implement the `Children_ToArrayAfter` which is like `ToArray` but starting with the node after the provided one.

Duckers commented on Source/Fuse.Controls.Panels/Layouts/WrapLayout.uno:
```diff
@@ -121,21 +121,22 @@ namespace Fuse.Layouts
 				clp.SetY(ItemHeight);
 
-			var placements = new float4[elements.Count];
+			var placements = new float4[elements.Children.Count];
```
My idea is to refactor this so only the Visual.Children.LinkedList.uno file uses the weird `Children_*` symbols. All other files should use a "proper" interface

Duckers commented on Source/Fuse.Nodes/Visual.Children.LinkedList.uno:
```diff
@@ -0,0 +1,216 @@
+			child.Children_parentID = -1;
+		}
+
+		internal Node Children_first;
```
As this is a mixin-like pattern, I use this naming convention to show that something fishy is going on. I'm thinking of this as a field of the `Children` mixin.

These symbols are not meant to be used outside of the file implementing the "mixin", I will refactor this to use proper properties instead.

Duckers commented on Source/Fuse.Nodes/Visual.Children.LinkedList.uno:
```diff
@@ -0,0 +1,216 @@
+			int i = 0;
+			while (k != null)
+			{
+				if (k == n) return i;
```
Why would a reference comparison be slower?

Comparing the ID would be faster, but only `Visual`s have those, not all nodes.

Duckers commented on Source/Fuse.Nodes/Visual.Children.LinkedList.uno:
```diff
@@ -0,0 +1,216 @@
+			else return Children_ItemAt(index-1);
+		}
+
+		void Children_Insert(int index, Node newNode)
```
It is still needed to implement `IList<T>`. Hopefully not used much in performance critical code paths, but must be there.

Duckers commented:
@mortoray I've addressed all your review points and also implemented a `SafeIterator` similar to what you suggested. As this in either case creates a new object to iterate, I simply went for implementing `IEnumerator<Node>` as a safe iterator (which performs lazy copy only if it has to).

Duckers commented:
Please re-review/approve.

kusma commented:
> Source\Fuse.Controls.Navigation\PageIndicator.uno(106.21): W4139: 'Fuse.Visual.get_LastVisualChild()' is obsolete: 'Use LastChild<Visual>() instead'

...and:

> From 31 projects, ran 911 tests, 134 failed.

kusma commented on Source/Fuse.Reactive.JavaScript/Tests/UX/ClassConstructor.ux:
```diff
@@ -1,6 +1,5 @@
 <UX.ClassConstructorBase ux:Class="UX.ClassConstructor">
 	<Panel ux:Dependency="A" ux:Binding="A" />
 	<Panel ux:Dependency="B" /> <!--  This one is here to force this and UX.ClassConstructorBase to get different constructor signatures -->
-	<Panel ux:Name="_a" Children="A" ux:AutoBind="false" />
```
You can't do this. This is a regression-test that verifies that we don't generate invalid Uno-code in this case, changing this completely undermines the test-case.

Duckers replied:
Fixed now

mortoray commented:
@Duckers How do this avoid creating a memory leak on the nodes? A list of children references prev/next, thus creating a memory loop. How is this loop cleaned up?

mortoray commented on Source/Fuse.Controls.Panels/Layouts/WrapLayout.uno:
```diff
@@ -121,21 +121,21 @@ namespace Fuse.Layouts
 			
-			for (int i = 0; i < elements.Count;++i)
+			int i = 0;
+			for (var e = container.FirstChild<Visual>(); e != null; e = e.NextSibling<Visual>(), i++)
```
The variable `i` has a different value than previously. You're only iterating `Visual` now instead of `Node`, thus `i++` is not incremented as often. You probably need to change this to `FirstChild<Node>` and `NextSibling<Node>` just to be safe.

mortoray commented on Source/Fuse.Controls.Panels/Layouts/WrapLayout.uno:
```diff
@@ -183,10 +183,9 @@ namespace Fuse.Layouts
 				var elp = lp.CloneAndDerive();
-				for (int i=0; i < elements.Count; ++i)
+				i = 0;
+				for (var element = container.FirstChild<Visual>(); element != null; element = element.NextSibling<Visual>(), i++)
```
Same here, not the same `i` value  (if this change was consistent it might work, but it feels safer to iteratre over `Node` like before).

mortoray commented on Source/Fuse.Controls/Tests/Fuse.Controls.Test.Helpers/PanelTester.uno:
```diff
@@ -122,6 +125,8 @@ namespace Fuse.Controls.Test
 					root.SnapToPixelsSize(float2(100, 71)), float2(800 - 100 - root.SnapToPixelsPos(21.4f),
 					root.SnapToPixelsPos(5.3f)));
+
+				root.Children.Remove(parent);
```
Why did you add this to test tests?

Duckers replied:
It is now illegal to be a child of more than more parent. This test violated that.

mortoray commented on Source/Fuse.Navigation/Navigation.Locators.uno:
```diff
@@ -26,11 +26,10 @@ namespace Fuse.Navigation
-				var c = node.Children[i] as IBaseNavigation;
-				//mean to check only behaviours (as was a distinct list before), so exclude visuals
-				if (c != null && !(c is Visual)) return c;
+				var c = x as IBaseNavigation;
```
Why not just do `node.FirstChild<IBaseNavigation>` in the for loop, or in place of the for-loop entirely?

Duckers replied:
The template argument must be `T : Node`, otherwise you can't call `.NextSibling<T>()`

mortoray replied:
The template argument doesn't need to be T to call `FirstChild<T>` though. It can be any type. It's only if you need iteration it needs to be a `Node`, but that doesn't affect `FirstChild`.  Actually, it doesn't even affect `NextSibling<T>` either. It's only when used together in a for-loop they need to be nodes, but that should just work since the for-loop knows the type.

I'm pointing this out since I know there's at least one case on master nor where `FindFirst` is expected to work with non-node types.

mortoray commented on Source/Fuse.Nodes/Node.DataContext.uno:
```diff
@@ -87,13 +87,13 @@ namespace Fuse
 					{
-						var sibdp = p.Children[i] as ISiblingDataProvider;
-						if (sibdp != null)
+						var sdp = dp as ISiblingDataProvider;
```
Is this not a `LastChild<ISiblingDataProvider>` loop instead of using a cast here?

Duckers replied:
The template argument must be T : Node, otherwise you can't call .NextSibling<T>()

mortoray commented on Source/Fuse.Nodes/Visual.Children.SafeIterator.uno:
```diff
@@ -0,0 +1,171 @@
+				if (_array != null)
+					return _pos < _array.Length;
+				
+				if (_current == null) _current = _v._firstChild;
```
I think this is technically wrong. If you call `MoveNext` two time as the end of a list this will result in it returning to the start of the list. It should instead continue to return `false` if you keep iterating.

Duckers replied:
Fixed in the code.

mortoray commented on Source/Fuse.Nodes/Visual.Children.SafeIterator.uno:
```diff
@@ -0,0 +1,171 @@
+				return _current != null;
+			}
+
+			public void Dispose()
```
And now you hit the Uno problem: https://github.com/fusetools/uno/issues/1148
Basically you can't rely on `Dispose` being called. I hit that prolem when trying to optimize an iterator as well.

Duckers replied:
@mortend  will address this shortly

mortoray commented on Source/Fuse.Nodes/Visual.Children.SafeIterator.uno:
```diff
@@ -0,0 +1,171 @@
+				}
+
+				if (_nextIterator != null) _nextIterator.SecureCopy();
+			}
```
Once you've done a `SecureCopy` you can remove `this` from the `_safeIterator` list. Since it has it's own copy there's no longer a need to track it as being an iterator.

mortoray commented on Source/Fuse.Nodes/Visual.Children.SafeIterator.uno:
```diff
@@ -0,0 +1,171 @@
+			{
+				if (_array == null)
+				{
+					if (MultipleIterators)
```
Perhaps check if there is a cachedArray instead, since there might be one even without multiple iterators.

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -18,35 +18,59 @@ namespace Fuse
+			return null;
+		}
 
-				i = Math.Min(i, Children.Count) - 1;
```
The reason the loop iterating backwards before, and did this check, was because removing one node could potentially cause additional nodes to be unrooted -- such as unrooting a behaviour that added children...

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -18,35 +18,59 @@ namespace Fuse
+			while (c != null)
+			{
+				var v = c as T;
+				c = c._nextSibling; // important to do this before calling Remove!
```
...by changing it to this you might actually be capturing a node that is no longer part of the children list.

I don't know of an easy solution other than using a SafeIterator here.

mortoray commented on Source/Fuse.Nodes/Visual.Children.uno:
```diff
@@ -219,27 +248,24 @@ namespace Fuse
-		void InsertNodesImpl(int index, IEnumerator<Node> items, bool allowMove)
+		void InsertNodesAfterImpl(Node preceeder, IEnumerator<Node> items, bool allowMove)
 		{
-			if (index <0 || index > Children.Count)
```
Perhaps replace this sanity check with one that checks that `preceeder` is a child of this node.  It's zero cost now I think, and it'd give a cleaner error than waiting until deep inside the loop.

Duckers replied:
Fixed in the code.

mortoray commented:
Once the tests pass and https://github.com/fusetools/uno/issues/1148 is fixed I can approve. I'd just like a chance to review any potential changes required to get the tests working.

Duckers commented:
@mortoray 1148 is merged to uno master now
